wicker/core/datasets.py

```diff
@@ -35,17 +94,127 @@ def __len__(self) -> int:
+                f.read().decode("utf-8"), treat_objects_as_bytes=self._treat_objects_as_bytes
+            )
+        return schema_data
 
     @abc.abstractmethod
     def arrow_table(self) -> pyarrow.Table:
         """Return the pyarrow table with all the metadata fields and pointers of the dataset."""
         pass
 
 
+class FileSystemDataset(AbstractDataset):
+    """Implementation of a Map-based dataset on local file system or mounted drive"""
+
+    def __init__(
+        self,
+        dataset_name: str,
+        dataset_partition_name: str,
+        dataset_version: str,
+        columns_to_load: Optional[List[str]] = None,
+        filelock_timeout_seconds: int = FILE_LOCK_TIMEOUT_SECONDS,
+        filesystem_root_path: Optional[str] = None,
+        local_cache_path_prefix: Optional[str] = os.getenv("TMPDIR", "/tmp"),
+        pa_filesystem: Optional[pafs.LocalFileSystem] = None,
+        path_factory: Optional[WickerPathFactory] = None,
+        storage: Optional[FileSystemDataStorage] = None,
+        treat_objects_as_bytes: bool = False,
+    ):
+        """Initializes a FileSystemDataset.
+
+        :param dataset_name: name of the dataset
+        :param dataset_partition_name: partition name
+        :param dataset_version: version of the dataset
+        :param columns_to_load: list of columns to load, defaults to None which loads all columns
+        :param filelock_timeout_seconds: number of seconds after which to timeout on waiting for downloads,
+            defaults to FILE_LOCK_TIMEOUT_SECONDS
+        :param filesystem_root_path: path to the root of the wicker file system. If path factory is none,
+            this must be set.
+        :param local_cache_path_prefix: Path to local cache path, if None don't create cache
+        :param pa_filesystem: Pyarrow filesystem for reading the parquet files and tables.
+        :param path_factory: Optional WickerPathFactory for pulling consistent paths.
+        :param storage: Optional FileSystemDataStorage object for pulling files from filesystem
+        :param treat_objects_as_bytes: If set, don't try to decode ObjectFields and keep them as binary data.
+        """
+        if path_factory is None and filesystem_root_path is None:
+            raise ValueError("Need to pass either path factory of wicker dataset or root of the tree.")
+        # ignore type failure here as we handle the case where they're both none above
+        path_factory = (
+            path_factory
+            if path_factory is not None
+            else WickerPathFactory(root_path=filesystem_root_path)  # type: ignore
+        )
+        pa_filesystem = pafs.LocalFileSystem() if pa_filesystem is None else pa_filesystem
+        storage = storage if storage is not None else FileSystemDataStorage()
+
+        super().__init__(
+            dataset_name=dataset_name,
+            dataset_partition_name=dataset_partition_name,
+            dataset_version=dataset_version,
+            storage=storage,
```

Comment:
Nit to sort the storage argument above after path_factory to match the order in which the parameters are declared in the `AbstractDataset` base class.